morello/src/search.rs

```diff
@@ -876,6 +867,221 @@ mod tests {
+            depth: 0,
+        }
+    }
+
+    #[test]
+    fn test_implreducer_sort_by_cost() {
+        let top_k = 3;
+        let preferences = smallvec![];
+        let mut reducer = ImplReducer::new(top_k, preferences);
+
+        let cost1 = create_simple_cost(1);
+        let cost2 = create_simple_cost(2);
+        let cost3 = create_simple_cost(3);
+
+        reducer.insert(0, cost1.clone());
+        reducer.insert(1, cost3.clone());
+        reducer.insert(2, cost2.clone());
+
+        let expected: SmallVec<[_; 1]> = smallvec![(0, cost1), (2, cost2), (1, cost3)];
+        assert_eq!(reducer.finalize(), expected);
+    }
+
+    #[test]
+    fn test_implreducer_sort_by_action_idx() {
+        let top_k = 3;
+        let preferences = smallvec![];
+        let mut reducer = ImplReducer::new(top_k, preferences);
+
+        let cost1 = create_simple_cost(1);
+
+        reducer.insert(1, cost1.clone());
+        reducer.insert(0, cost1.clone());
+        reducer.insert(2, cost1.clone());
+
+        let expected: SmallVec<[_; 1]> =
+            smallvec![(0, cost1.clone()), (1, cost1.clone()), (2, cost1.clone())];
+        assert_eq!(reducer.finalize(), expected);
+    }
+
+    #[test]
+    fn test_implreducer_sort_by_cost_then_action_idx() {
+        let top_k = 3;
+        let preferences = smallvec![];
+        let mut reducer = ImplReducer::new(top_k, preferences);
+
+        let cost1 = create_simple_cost(1);
+        let cost2 = create_simple_cost(2);
+
+        reducer.insert(1, cost1.clone());
+        reducer.insert(0, cost2.clone());
+        reducer.insert(2, cost1.clone());
+
+        let expected: SmallVec<[_; 1]> =
+            smallvec![(1, cost1.clone()), (2, cost1.clone()), (0, cost2)];
+        assert_eq!(reducer.finalize(), expected);
+    }
+
+    #[test]
+    fn test_implreducer_preference_replacement() {
+        let top_k = 3;
+        let preferences = smallvec![0, 3, 3];
+        let mut reducer = ImplReducer::new(top_k, preferences);
+
+        let cost1 = create_simple_cost(1);
+
+        reducer.insert(0, cost1.clone());
+        reducer.insert(1, cost1.clone());
+        reducer.insert(2, cost1.clone());
+        reducer.insert(3, cost1.clone());
+
+        let expected: SmallVec<[_; 1]> =
+            smallvec![(0, cost1.clone()), (2, cost1.clone()), (3, cost1)];
+        assert_eq!(reducer.finalize(), expected);
+    }
+
+    #[test]
+    fn test_implreducer_preference_replacement2() {
+        let top_k = 3;
+        let preferences = smallvec![0, 0, 3];
+        let mut reducer = ImplReducer::new(top_k, preferences);
+
+        let cost1 = create_simple_cost(1);
+
+        reducer.insert(0, cost1.clone());
+        reducer.insert(1, cost1.clone());
+        reducer.insert(2, cost1.clone());
+        reducer.insert(3, cost1.clone());
+
+        let expected: SmallVec<[_; 1]> =
+            smallvec![(0, cost1.clone()), (1, cost1.clone()), (3, cost1)];
+        assert_eq!(reducer.finalize(), expected);
+    }
```

Comment:
Yes, preferences are 0, 0, 3 (instead of 0, 3, 3) and the result would be 0, 1, 3 (instead of 0, 2, 3). But as you mentioned, preferences are not supposed to hold duplicates, so this test should be tweaked.